src/Mod/PartDesign/App/FeatureHole.cpp

```diff
@@ -1882,30 +1886,72 @@ App::DocumentObjectExecReturn* Hole::execute()
             // we reuse the name protoHole (only now it is threaded)
             protoHole = mkFuse.Shape();
         }
+        std::vector<TopoShape> holes;
+        auto compound = findHoles(holes, profileshape, protoHole);
 
-        TopoDS_Compound holes = findHoles(profileshape.getShape(), protoHole);
-        this->AddSubShape.setValue(holes);
+        TopoShape result(0);
 
-        // For some reason it is faster to do the cut through a BooleanOperation.
-        BRepAlgoAPI_Cut mkBool(base.getShape(), holes);
-        if (!mkBool.IsDone()) {
-            return new App::DocumentObjectExecReturn(QT_TRANSLATE_NOOP("Exception", "Boolean operation failed"));
-        }
-        TopoDS_Shape result = mkBool.Shape();
+        // set the subtractive shape property for later usage in e.g. pattern
+        this->AddSubShape.setValue(compound);
 
+        if (base.isNull()) {
+            Shape.setValue(compound);
+            return App::DocumentObject::StdReturn;
+        }
 
-        // We have to get the solids (fuse sometimes creates compounds)
-        base = getSolid(result);
-        if (base.isNull())
-            return new App::DocumentObjectExecReturn(QT_TRANSLATE_NOOP("Exception", "Resulting shape is not a solid"));
-        base = refineShapeIfActive(base);
+        // First try cutting with compound which will be faster as it is done in
+        // parallel
+        bool retry = true;
+        const char *maker;
+        switch (getAddSubType()) {
+            case Additive:
+                maker = Part::OpCodes::Fuse;
+                break;
+            default:
+                maker = Part::OpCodes::Cut;
+        }
+        try {
+            if (base.isNull())
+                result = compound;
+            else
+                result.makeElementBoolean(maker, {base,compound});
+            result = getSolid(result);
+            retry = false;
+        } catch (Standard_Failure & e) {
+            FC_WARN(getFullName() << ": boolean operation with compound failed ("
+                                  << e.GetMessageString() << "), retry...");
+        } catch (Base::Exception & e)  {
+            FC_WARN(getFullName() << ": boolean operation with compound failed ("
+                                  << e.what() << "), retry...");
+        }
 
-        if (!isSingleSolidRuleSatisfied(base.getShape())) {
-            return new App::DocumentObjectExecReturn(
-                QT_TRANSLATE_NOOP("Exception", "Result has multiple solids: that is not currently supported."));

```

Comment:
I think we can keep this part of the code, guess it will be useful in the future developments of MultiSolid modeling

Reply:
Good catch, I was working from the incoming code and missed the feature add in the existing code.  Fixed.